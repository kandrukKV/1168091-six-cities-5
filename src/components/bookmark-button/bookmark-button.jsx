import React, {PureComponent} from "react";
import {redirectToRouteAction} from "../../store/action";
import {connect} from "react-redux";
import {AuthorizationStatus, APIRoute, FavoriteStatus} from "../../const";
import {getAuthorizationStatusSelector} from "../../store/selectors";
import {setOfferStatus} from "../../store/api-actions";
import PropTypes from "prop-types";

class BookmarkButton extends PureComponent {

  constructor(props) {
    super(props);
    this.onButtonClickHandler = this.onButtonClickHandler.bind(this);
  }

  onButtonClickHandler() {
    const {isFavorite, authorizationStatus, cardId, redirectToLoginPage, changeOfferStatus} = this.props;

    switch (authorizationStatus) {
      case AuthorizationStatus.NO_AUTH:
        return redirectToLoginPage;
      case AuthorizationStatus.AUTH:
        return () => {
          return changeOfferStatus(cardId, isFavorite ? FavoriteStatus.NO : FavoriteStatus.YES);
        };
    }
    return () => {};
  }

  render() {
    const {
      className = `place-card`,
      buttonWidth = 18,
      buttonHeight = 19,
      isFavorite
    } = this.props;

    const buttonClassActive = isFavorite ? `${className}__bookmark-button--active` : ``;

    return (
      <button
        onClick={this.onButtonClickHandler()}
        className={`${className}__bookmark-button ${buttonClassActive} button`}
        type="button"
      >
        <svg className={`place-card__bookmark-icon`} width={buttonWidth} height={buttonHeight}>
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }
}

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatusSelector(state)
});

const mapDispatchToProps = (dispatch) => ({
  redirectToLoginPage() {
    dispatch(redirectToRouteAction(APIRoute.LOGIN));
  },
  changeOfferStatus(hotelId, status) {
    dispatch(setOfferStatus(hotelId, status));
  }
});

BookmarkButton.propTypes = {
  isFavorite: PropTypes.bool.isRequired,
  className: PropTypes.string,
  buttonWidth: PropTypes.number,
  buttonHeight: PropTypes.number,
  authorizationStatus: PropTypes.string.isRequired,
  cardId: PropTypes.number.isRequired,
  redirectToLoginPage: PropTypes.func.isRequired,
  changeOfferStatus: PropTypes.func.isRequired,
};

export {BookmarkButton};
export default connect(mapStateToProps, mapDispatchToProps)(BookmarkButton);
