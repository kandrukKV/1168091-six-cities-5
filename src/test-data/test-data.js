export const noon = () => {};

export const additions = [
  `addition1`,
  `addition2`,
  `addition3`,
  `addition4`,
  `addition5`,
];

export const cities = [
  `Amsterdam`,
  `Paris`,
];

export const mockOffers = [
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13
      }
    },
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    ],
    title: `Perfectly located Castro`,
    isFavorite: false,
    isPremium: false,
    rating: 1,
    type: `house`,
    bedrooms: 3,
    adults: 7,
    price: 440,
    additions: [
      `Dishwasher`,
      `Dishwasher`,
      `Dishwasher`,
    ],
    host: {
      id: 25,
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`
    },
    description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    location: {
      latitude: 52.385540000000006,
      longitude: 4.902976,
      zoom: 16
    },
    id: 1
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13
      }
    },
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    ],
    title: `Perfectly located Castro`,
    isFavorite: true,
    isPremium: true,
    rating: 2,
    type: `house`,
    bedrooms: 3,
    adults: 7,
    price: 440,
    additions: [
      `Dishwasher`,
      `Dishwasher`,
      `Dishwasher`,
    ],
    host: {
      id: 25,
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`
    },
    description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    location: {
      latitude: 52.388540000000006,
      longitude: 4.899976,
      zoom: 16
    },
    id: 2
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13
      }
    },
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    ],
    title: `Perfectly located Castro`,
    isFavorite: false,
    isPremium: true,
    rating: 3,
    type: `house`,
    bedrooms: 3,
    adults: 7,
    price: 440,
    additions: [
      `Dishwasher`,
      `Dishwasher`,
      `Dishwasher`,
    ],
    host: {
      id: 25,
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`
    },
    description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    location: {
      latitude: 52.364540000000005,
      longitude: 4.9019759999999994,
      zoom: 16
    },
    id: 3
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13
      }
    },
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    ],
    title: `Perfectly located Castro`,
    isFavorite: true,
    isPremium: false,
    rating: 4,
    type: `house`,
    bedrooms: 3,
    adults: 7,
    price: 440,
    additions: [
      `Dishwasher`,
      `Dishwasher`,
      `Dishwasher`,
    ],
    host: {
      id: 25,
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`
    },
    description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    location: {
      latitude: 52.385540000000006,
      longitude: 4.886976,
      zoom: 16
    },
    id: 4
  },
  {
    city: {
      name: `Amsterdam`,
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13
      }
    },
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    images: [
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
      `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    ],
    title: `Perfectly located Castro`,
    isFavorite: true,
    isPremium: true,
    rating: 5,
    type: `house`,
    bedrooms: 3,
    adults: 7,
    price: 440,
    additions: [
      `Dishwasher`,
      `Dishwasher`,
      `Dishwasher`,
    ],
    host: {
      id: 25,
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`
    },
    description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
    location: {
      latitude: 52.397540000000006,
      longitude: 4.9099759999999995,
      zoom: 16
    },
    id: 5
  }
];

export const mosckReviews = [
  {
    id: 1,
    user: {
      id: 18,
      name: `Sophie`,
      avatar: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/9.jpg`
    },
    rating: 3,
    comment: `We loved it so much, the house, the veiw, the location just great.. Highly reccomend :)`,
    date: `2020-11-02T13:38:44.752Z`
  },
  {
    id: 2,
    user: {
      id: 16,
      name: `Mollie`,
      avatar: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/7.jpg`
    },
    rating: 2,
    comment: `What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!`,
    date: `2020-11-02T13:38:44.752Z`
  }
];
