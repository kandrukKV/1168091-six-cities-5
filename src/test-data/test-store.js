import configureStore from "redux-mock-store";
import {mockOffers, mosckReviews, cities} from "../test-data/test-data";

const createStore = configureStore();

const mockStore = createStore({
  API_DATA: {
    offers: mockOffers,
    favoriteOffers: mockOffers,
    offerDetails: {
      card: mockOffers[0],
      reviews: mosckReviews,
      nearPlaces: mockOffers
    }
  },
  APP_STATE: {
    currentCity: `Amsterdam`,
    cities,
    currentSortType: `Popular`,
    activeCard: null,
    reviewFormState: `EDITING`
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    authInfo: null
  }
});

export {mockStore};
