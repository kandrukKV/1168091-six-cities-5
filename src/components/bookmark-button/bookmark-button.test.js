import React from "react";
import renderer from "react-test-renderer";
import {BookmarkButton} from "./bookmark-button";
import {noon} from "../../test-data/test-data";

describe(`BookmarkButton render correctly`, () => {
  it(`BookmarkButton is favorite`, () => {
    const tree = renderer.create(
        <BookmarkButton
          isFavorite={true}
          authorizationStatus={`NO_AUTH`}
          cardId={1}
          buttonWidth={22}
          buttonHeight={33}
          className={`new_class`}
          redirectToLoginPage={noon}
          changeOfferStatus={noon}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`BookmarkButton is't favorite`, () => {
    const tree = renderer.create(
        <BookmarkButton
          isFavorite={false}
          authorizationStatus={`AUTH`}
          cardId={1}
          buttonWidth={22}
          buttonHeight={33}
          className={`new_class`}
          redirectToLoginPage={noon}
          changeOfferStatus={noon}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`BookmarkButton with default values`, () => {
    const tree = renderer.create(
        <BookmarkButton
          isFavorite={false}
          authorizationStatus={`AUTH`}
          cardId={1}
          redirectToLoginPage={noon}
          changeOfferStatus={noon}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
