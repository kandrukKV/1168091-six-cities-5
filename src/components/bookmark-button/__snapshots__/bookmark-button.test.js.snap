// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkButton render correctly BookmarkButton is favorite 1`] = `
<button
  className="new_class__bookmark-button new_class__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={33}
    width={22}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`BookmarkButton render correctly BookmarkButton is't favorite 1`] = `
<button
  className="new_class__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={33}
    width={22}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`BookmarkButton render correctly BookmarkButton with default values 1`] = `
<button
  className="place-card__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
